rfc/0075-faster-failover-and-configuration-push.md

```diff
@@ -0,0 +1,364 @@
+`DedupeNotMyVbucketClustermap` option is negotiated.
+
+It is not guaranteed that the configuration will be immediately available on the all nodes of the cluster once one of
+the node sends `CLUSTERMAP_CHANGE_NOTIFICATION` request to SDK. To workaround this issue, the SDK should remeber
+`epoch`/`revision` pair from notification payload as "recently announced" version, and ensure that the configuration
+monitor will continue attempts to fetch configuration until the received body will have version that is not older than
+"recently announced".
+
+As an optional optimization, the SDK might perform attempts to retrieve configuration starting from the node that
+pushed configuration notification. Although it might not be possible to do easily in all SDKs.
+
+#### Mixed Clusters
+
+In clusters where there is a mix of nodes with older server versions, meaning that some nodes do not acknowledge
+`ClustermapChangeNotificationBrief`, the respective connection should notify the configuration monitor about its lack of
+support for configuration pushes from the server. As a result, the monitor should utilize the old polling mechanism for
+this particular node instead.
+
+In other words, the configuration monitor should continue polling only if there are nodes, that do not support
+`ClustermapChangeNotificationBrief`, otherwise configuration monitor issue `GET_CLUSTER_CONFIG` request only when
+notification arrives.
+
+### Bootstrap Changes
+
+[RFC-0048][rfc-0048] describes bootstrap process for the SDK and the KV connections in particular. With Faster Failover
+mechanism implemented, the SDK should not start polling for the nodes, where the KV engine has acknowledged
+`ClustermapChangeNotificationBrief` feature. Such connections are expected to be notified by the KV engine when the
+configuration will be received.t
+
+### Enhancements in Handling the `NotMyVbucket` Status
+
+`DedupeNotMyVbucketClustermap` feature allows to save traffic by not sending configuration, if SDK already seen the same
+revision. 
+
+Several modifications are required in the SDK:
+1. Response handler should tolerate empty response with `NotMyVbucket` status, as the KV engine assumes that the SDK
+   already seen configuration, and there is no newer configuration available. In this case SDK should just retry the
+   operation.
+
+2. If the reponse payload contains body, it contains current configuration, which should be sent to configuration
+   monitor (manager). The SDK should either synchronously apply configuration, create waiting queue for given
+   `epoch`/`revision` pair.
+   Once configuration applied, the SDK must check if new configuration routes the operation to new endpoint or new
+   vbucket on the old endpoint, and *immediately* dispatch operation to new endpoint (or same endpoint in case vbucketID
+   has changed). In any other case, the SDK should send operation to retry orchestrator.
```

Comment:
I'm not sure I get this part.  Say the SDK is doing 10,000 operations per second per node.  With dedup, just one of those operations will get a NVMB with a config - the other 9,999 will get a NVMB with an empty body.  Why is there a bunch of special case logic for the single operation?  It would be a _lot_ simpler to implement in the JVM clients if we could asynchronously apply the NVMB config body (all config updating is async in JVM world), and send all requests into the RetryOrchestrator but on a tight retry - like a few milliseconds.  That would be pretty easy to implement and those few milliseconds wouldn't seem to impact fast failover at all.

Reply:
@programmatix How would you provide determinism that the config had been applies before you attempted to execute the operation again?

Reply:
I haven't pathfound in the JVM clients yet (David's assigned but asked me to look at it while he's away), but IIRC the config mechanism is an async pub-sub.  Something pushes a config and all subscribers will, unless something has gone tragically wrong, get that config in nanoseconds.  So I'm proposing - and I may want to revise this after pathfinding - permitting an alternative setup that's technically non-deterministic, but will be in practice in basically all cases: send the new config to the config publisher, and scheduled the operation to be retried in X milliseconds.  It should be easy to implement, and it's much less risk than introducing a new blocking mechanism to the config publisher that will be challenging to test and will only trigger during rebalances.  And if something odd happens and the config publisher hasn't published by X milliseconds, then all that happens is operations are retried, we get another NVMB, and go around again.  That would be some more traffic to the cluster - but to stress, this is an in-memory pub-sub stream, and it should always have delivered the config within those X millis.

Reply:
I agree with Graham's commentary. "Immediately retry now" isn't feasible for the JVM SDKs.